Add ElectronicEigenvalues and ElectronicBandStructure properties

JosePizarro3 opened the pull request:
This pr contains the definition for the former `ElectronicBandStructure` with the new schema (`PhysicalProperties` and `Variables`), as well as the new property `ElectronicEigenvalues`. During development, I also wanted to define `FermiSurface`, but this makes more sense in a separate pr (#78).

Some important changes with respect to the old schema and normalization:
- Now, there is a `value_contributions` whose objective is to group all the former stuff that was in `BandEnergies` as `value_<name-of-the-contribution>`
- Now there is not a concept of segments. I rather changed for storing `KLinePath(Variables).points` as a flat list of points
- I completely restructured the old `add_band_gap` normalization. Now, we have `extract_band_gap` and, mainly, `resolve_homo_lumo_eigenvalues` to handle this.
- The old `get_special_points` function is re-defined under `numerical_settings.py:KMesh`. This was a very buggy function, specially when the path is not defined for points related by symmetry (e.g., `'X'` is both `[0.5, 0, 0]` and `[-0.5, 0, 0]`. Now we aim people to define the strings related with the high symmetry points OR manually define the path.

One question is whether `reciprocal_cell` will be still here in this schema or we can resolve it in another way. This quantity is only interesting for the front end.

JosePizarro3 committed:
Improved numerical_settings schema for KSpace

Added KLinePath and KMesh variables with refs to the NumericalSettings sections

JosePizarro3 committed:
Definitely, git is trash sometimes. I hate when rebasing and it gets completely crazy....... I had to squash all the shit into one single commit. It was that, or just killing myself.

JosePizarro3 committed:
Restore .vscode/settings.json file

Deleted unused comment-warning in variables.py

JosePizarro3 committed:
Fix issue when squashing and deleting generate_electronic_eigenvalues from conftest

Fix bulk type in resolve_reciprocal_lattice_vectors

JosePizarro3 committed:
Added reciprocal_cell, method to resolve it and testing for ElectronicEigenvalues

JosePizarro3 committed:
Fix testing and schema to be QuantityReference

JosePizarro3 committed:
Fixed testing

JosePizarro3 commented:
Hey @ladinesa this is the pr I told you about.

The focus should be the changes in the new module, `src/nomad_simulations/properties/band_structure.py`. I still have some todos, but I moved them to separated issues (#77 #78 #79).

Thanks!

JosePizarro3 committed:
Patch over self.rank so that msgpck does not get crazy with numpy.int32 (it needs int types)

ladinesa commented on src/nomad_simulations/properties/band_structure.py:
```diff
@@ -0,0 +1,370 @@
+    ) -> None:
+        super().__init__(m_def, m_context, **kwargs)
+        # ! `n_bands` need to be set up during initialization of the class
+        self.rank = [self.n_bands]
```
or maybe get it from kwargs

JosePizarro3 replied:
Yeah, sorry, at this point I do not care too much about this; at the end, this is a definition, so it should go in m_def. 

I am waiting for Markus to implement something on this direction, so my point is: don't pay too much attention to this dirty assignment for now 😄

ladinesa replied:
I was just concerned about the same error you were having previously that you have to convert it to int when read from msection

JosePizarro3 replied:
Yes, I had to, thanks for catching it up. In the new version you can see I did `self.rank = [int(self.n_bands)]`, but I will change it 👍🏻

ladinesa commented on src/nomad_simulations/properties/band_structure.py:
```diff
@@ -0,0 +1,370 @@
+        super().__init__(m_def, m_context, **kwargs)
+        self.name = self.m_def.name
+
+    def _check_occupation_shape(self, logger: BoundLogger) -> bool:
```
maybe rename this to is_valid_occupation

JosePizarro3 replied:
I used `_check` for this, but indeed, `is_valid` sounds better as a validation mechanism. I will change it in other modules as well.

JosePizarro3 replied:
Ok, I went for `validate_` instead.

ladinesa commented on src/nomad_simulations/properties/band_structure.py:
```diff
@@ -0,0 +1,370 @@
+        super().__init__(m_def, m_context, **kwargs)
+        self.name = self.m_def.name
+
+    def _check_occupation_shape(self, logger: BoundLogger) -> bool:
```
not sure why you are passing logger? i would define it globally LOGGER = utils.get_logger(__name__)

JosePizarro3 replied:
Yes, this is confusing for me. As I am passing logger to `normalize()` would these conflict? I thought that I had to keep the same `logger` in all places to write the messages in the same bound logger.

ladinesa replied:
ah yeah right then ignore my comment.

ladinesa commented on src/nomad_simulations/properties/band_structure.py:
```diff
@@ -0,0 +1,370 @@
+        # Sorting `value` and `occupation`
+        sorted_value, sorted_occupation = self.order_eigenvalues(logger)
+        if sorted_value is None or sorted_occupation is None:
+            if self.highest_occupied is not None and self.lowest_unoccupied is not None:
```
are you not supposed to return right away self.highest(lowest)(un)occupied if they are assigned? or maybe check if the values are consistent.

JosePizarro3 replied:
Good point, but my argument is that normalization is more important here than parsing, i.e., if sorted_value and sorted_occupation could be extracted, then we used those (and not the parsed data) for extracting `highest_occupied` and `lowest_unoccupied`. Otherwise, we return then if they exist.

ladinesa commented on src/nomad_simulations/properties/band_structure.py:
```diff
@@ -0,0 +1,370 @@
+        lumo = self.lowest_unoccupied
+        low_occ = 0
+        high_unocc = sorted_value.shape[-1] - 1
+        while low_occ <= high_unocc:
```
can't you use np.argwhere(sorted_occupation > tol) here? also, how about spin-pol case?

JosePizarro3 replied:
Hehe, I'll argue that binary search (which is logN) is more optimal than argwhere (which is linear). These arrays are going to be potentially large, but I can change it to argwhere if even tho you prefer readability.

ladinesa replied:
there is also this np.argmax that seems to return the first element matching criterion.

JosePizarro3 replied:
It seems to be linear as well. ChatGPT helped me check this (for a sample of 10000 points)

![time_complexity](https://github.com/nomad-coe/nomad-schema-plugin-simulation-data/assets/112697669/993695ba-e2b7-42a4-8de3-3e9b34a1ee2b)

ladinesa replied:
interesting. thanks for testing. then go for the optimized approach.

JosePizarro3 replied:
I just found `numpy.searchsorted` which seems to perform binary search. Let me try it.

JosePizarro3 replied:
It seems to be almost the same. I tried some random arrays (hence the bit of randomness between results), averaged, and increased the size to 1e6.

![time_complexity](https://github.com/nomad-coe/nomad-schema-plugin-simulation-data/assets/112697669/802045a0-95e4-4598-bf98-d4b4cd2720aa)

Well, I can also use numpy.searchsorted if it is more readable, I don't mind actually.

ladinesa replied:
I suggest you use numpy for easy maintenance.

JosePizarro3 committed:
Added missing spin_channel to ElectronicEigenvalues

ladinesa commented on src/nomad_simulations/properties/band_structure.py:
```diff
@@ -0,0 +1,370 @@
+            (Optional[FermiSurface]): The extracted Fermi surface section to be stored in `Outputs`.
+        """
+        # Check if the system has a finite band gap
+        homo, lumo = self.resolve_homo_lumo_eigenvalues(logger)
```
why not access the homo and lumo from msection instead of runnung the function again?

JosePizarro3 replied:
I thought about it, but what if someone wants to `extract_band_gap` or `extract_fermi_surface` when homo and lumo are not set up and without normalizing first?

ladinesa replied:
ah my bad, the function already has a check for the homo/lumo values in the archive.

JosePizarro3 replied:
no problem, indeed it checks this

ladinesa commented:
Looks all fine to me except for some minor improvements. Do I check also the tests?

JosePizarro3 commented:
> Looks all fine to me except for some minor improvements. Do I check also the tests?

Thanks! I think the tests are just there to help understand the code, so up to you. Coverage is 99% (whatever that means 😅), so I guess I did a good job.

JosePizarro3 committed:
Added Alvin comments

JosePizarro3 committed:
Changed is_valid and _check methods for validate_

ndaelman-hu commented on src/nomad_simulations/numerical_settings.py:
```diff
@@ -803,8 +803,8 @@ def resolve_reciprocal_lattice_vectors(
             if is_not_representative(model_system, logger):
                 continue
-            # TODO extend this for other dimensions (@ndaelman-hu)
-            if model_system.type != 'bulk':
```
Don't remove this TODO, pls.

ndaelman-hu commented on .vscode/settings.json:
Settings should be in `.gitignore`. Nobody wants their VSCode setup messed up because they opened a project.

ndaelman-hu commented on src/nomad_simulations/numerical_settings.py:
```diff
@@ -165,15 +165,15 @@ class KSpaceFunctionalities:
     """
 
-    def _check_reciprocal_lattice_vectors(
+    def validate_reciprocal_lattice_vectors(
```
So, the name and visibility change are all fine.
I'd just ask to keep these for separate PRs (don't even have to be reviewed) as much as possible, as they distract from the important code changes at hand.

JosePizarro3 replied:
indeed, I just tackled the name change at once, but you are right

ndaelman-hu commented on src/nomad_simulations/properties/band_structure.py:
```diff
@@ -0,0 +1,377 @@
+class BaseElectronicEigenvalues(PhysicalProperty):
+    """
+    A base section used to define basic quantities for the `ElectronicEigenvalues`, `FermiSurface`, and
+    `ElectronicBandStructure` properties. This section serves to define `FermiSurface` and without needing to specify
+    other quantities that appear in `ElectronicEigenvalues`
```
> This section serves to define `FermiSurface` ~~and~~ without needing to

I'd even shorten it to: "This section defines common quantities between `FermiSurface` and `ElectronicBandStructure`.

JosePizarro3 replied:
I've changed the description, as strictly speaking `FermiSurface` as a property contains different values (not the energies, but the actual k points $\vec{k}_F$)

ndaelman-hu commented on src/nomad_simulations/properties/band_structure.py:
```diff
@@ -0,0 +1,377 @@
+        type=np.float64,
+        shape=['*', 'n_bands'],
+        description="""
+        Occupation of the electronic eigenvalues. This is a number between 0 and 2, where 0 means
+        that the state is unoccupied and 2 means that the state is fully occupied. It is controlled
+        by the Fermi-Dirac distribution:
+
+            $ f(E) = 1 / (1 + exp((E - E_F) / kT)) $
+
+        The shape of this quantity is defined as `[KMesh.n_points, KMesh.dimensionality, n_bands]`, where `KMesh` is a `variable`.
```
1. Since you seem to be handling spin-states, shouldn't it be restricted to 1? I guess what you want, is that the max filling depends on whether spin_channel is set, right? Either elaborate in the description OR add a quantity `max_occupation`. The latter would simplify the code logic.
2. I understand what you want to say with the line on the Fermi-Dirac distribution, but it comes across as confusing, e.g. the band structure also depends on other factors, wha about strongly-correlated systems, etc.  Most of our band structure don't have any T-dependence, so I wouldn't bother mentioning this. For those that do, it'll be the calculation that dictates the data.

JosePizarro3 replied:
Indeed, you are right, I totally overlooked the spin. I added `spin_channel` and I will adapt the description.

I will also adapt the Fermi-Dirac distribution. However, for a system of fermions, the Fermi-Dirac statistics is a necessary condition, independently of whether they are correlated or not; this is what is behind of certain filling conditions, e.g. the [Luttinger's theorem](https://en.wikipedia.org/wiki/Luttinger%27s_theorem), and it only depends on whether the system is a Fermi liquid or not, hence whether we can apply Fermi statistics or not. The only situations where these does not apply is near quantum criticality and non-Fermi liquid behaviors (I mean, the name says it everything 😄)$^{[1]}$.


$^{[1]}$ I was trying to find a good reference for this, but I couldn't. The most I can share with you is some of the initial chapters of Piers Coleman's book, [Introduction to Many Body Physics](https://www.cambridge.org/core/books/introduction-to-manybody-physics/B7598FC1FCEE0285F5EC767E835854C8). Let me know if you want the book.

ndaelman-hu replied:
I'll take a look. I thought that correlation would be a factor in deviating from the Fermi liquid model, but I'll read up some more.

ndaelman-hu commented on src/nomad_simulations/properties/band_structure.py:
```diff
@@ -0,0 +1,377 @@
+        """,
+    )
+
+    highest_occupied = Quantity(
+        type=np.float64,
+        unit='joule',
+        description="""
+        Highest occupied electronic eigenvalue. Together with `lowest_unoccupied`, it defines the
+        electronic band gap.
+        """,
+    )
+
+    lowest_unoccupied = Quantity(
+        type=np.float64,
+        unit='joule',
+        description="""
+        Lowest unoccupied electronic eigenvalue. Together with `highest_occupied`, it defines the
+        electronic band gap.
+        """,
+    )
```
As an aside, do you see any value in reporting the min/max of the eigenvalue sampling itself?
Most calculations don't cover the core states.

JosePizarro3 replied:
I think it would be interesting to cover min/maxs of specific set of bands, e.g., around the Fermi level, higher energy bands separated by gaps, and equivalently, core bands separated by gaps.

I will add a question to see whether we can do this later along the road. Thanks for bringing it up.

ndaelman-hu commented on src/nomad_simulations/properties/band_structure.py:
```diff
@@ -0,0 +1,377 @@
+        """,
+    )
+
+    reciprocal_cell = Quantity(
+        type=KSpace.reciprocal_lattice_vectors,
+        description="""
+        Reference to the reciprocal lattice vectors stored under `KSpace`. This reference is useful when resolving the Brillouin zone
+        for the front-end visualization.
+        """,
+    )
```
If you say that this is for visualization, this hould really only exist in `results`.
Personally, I think it can have a use here for other post-processing.

JosePizarro3 replied:
Ok, I delete the phrase referring to the front-end visualization.

ndaelman-hu commented on src/nomad_simulations/properties/band_structure.py:
```diff
@@ -0,0 +1,377 @@
+        super().__init__(m_def, m_context, **kwargs)
+        self.name = self.m_def.name
+
+    def validate_occupation(self, logger: BoundLogger) -> bool:
+        """
+        Validate `occupation` by checking if they exist and have the same shape as `value`.
+
+        Args:
+            logger (BoundLogger): The logger to log messages.
+
+        Returns:
+            (bool): True if the shape of `occupation` is the same as `value`, False otherwise.
+        """
+        if self.occupation is None or len(self.occupation) == 0:
+            logger.warning('Cannot find `occupation` defined.')
+            return False
+        if self.value is not None and self.value.shape != self.occupation.shape:
+            logger.warning(
+                'The shape of `value` and `occupation` are different. They should have the same shape.'
+            )
+            return False
+        return True
```
Just add `shape` to the definition of `value`. If you have to override it later on, that easy with `m_copy`.
Also, I'd use plural for quantities that store arrays.

JosePizarro3 replied:
There are other things going on with `value` and its `__setattr__` defined in `PhysicalProperty`, so defining `shape` will not have any effect.

I just want to double check that `occupation` and `value` has the same shape, but it is true that this could maybe be a decorator, as it can become a quite popular check?

ndaelman-hu replied:
I think a lot of our checks could be recast into decorators. I was already looking into it, but I'd have to gather all use cases. This is cleaning up for later.

ndaelman-hu commented on src/nomad_simulations/properties/band_structure.py:
```diff
@@ -0,0 +1,377 @@
+
+    def order_eigenvalues(
+        self, logger: BoundLogger
+    ) -> Tuple[Optional[pint.Quantity], Optional[np.ndarray]]:
```
No need for `Optional` here, it'll only complicate the logic down the line. Just use empty arrays to denote "failure". It'll keep the logic simpler.

JosePizarro3 replied:
Problem comes with the units of the returned `sorted_value`.

ndaelman-hu replied:
But, empty lists can also have units...

JosePizarro3 replied:
Yeah, but knowing the unit implies that the extraction works. I think None and Optional is much more clean. I also doubt that an empty pint list is not screwing things on the archive, meaning, that the quantity is set to an empty value instead of being None, which is the behavior we want

ndaelman-hu replied:
> I also doubt that an empty pint list is not screwing things on the archive, meaning, that the quantity is set to an empty value instead of being None, which is the behavior we want

That would be good to know, actually! Let's check it out. Tbh, the archive should have a way of handling empy arrays.

ndaelman-hu replied:
> I think None and Optional is much more clean.

They are definitely more verbose, which indeed can be more communicative.
I'd call simpler logic _cleaner_, though. If a type (e.g. `list`, `int`) can handle "failure" (e.g. `[]`, `0`) by itself, I wouldn't complicate matters.
This is, of course, very situation-dependent, as sometimes "failure" may present itself differently, e.g. `1 / 0`.

Honestly, Python's way of handling `None` is pretty naive. In a language like Haskell I'd make less of an issue, as the logic is better separated. There, `Nothing` is a constructor of the **monad** `Maybe`, which handles "failure" propogation in a more automated and modular fashion.

ndaelman-hu commented on src/nomad_simulations/properties/band_structure.py:
```diff
@@ -0,0 +1,377 @@
+            (Tuple[Optional[pint.Quantity], Optional[list]]): The flattened and sorted `value` and `occupation`.
+        """
+
+        # Check if `value` exists
+        if self.value is None or len(self.value) == 0:
+            logger.error('Could not find `value` defined.')
+            return None, None
```
Like here: this teste isn't necessary if you maintain arrays as the output.
Moreover, shouldn't instantiation assure that only 1 condition has to be met?

JosePizarro3 replied:
Well, someone could call for this method without `value` set up. Now, for the `is None` and `len()`, it is true that the length check does not make sense, so I will remove it.

ndaelman-hu commented on src/nomad_simulations/properties/band_structure.py:
```diff
@@ -0,0 +1,377 @@
+            logger.error('Could not find `value` defined.')
+            return None, None
+
+        # Check if `occupation` exists and have the same shape as `value`
```
typo in the conjugation

JosePizarro3 committed:
Added validate_quantity_wrt_value decorator for PhysicalProperty

Fixed test_band_structure

Fixed methods in ElectronicEigenvalues

Added .vscode/settings.json to gitignore

JosePizarro3 committed:
Added testing for decorator validate_quantity_wrt_value

JosePizarro3 committed:
Fix mypy: yes, it is crap sometimes

JosePizarro3 committed:
Implemented np.searchsorted instead of custom binary search in resolve_homo_lumo_eigenvalues

JosePizarro3 commented:
Ok, this pr is ready. I added also a validator for comparing a quantity with respect to `value` in any PhysicalProperty, to ensure consistence in all the cases where we don't want to consider a quantity a physical property, but to be actually under an existing one.

Thanks for the review @ladinesa @ndaelman-hu ! 🥳
